### PR TITLE
Lazy shares

### PR DESCRIPTION
**Pull request by rullzer**
For https://github.com/owncloud/core/pull/20571#issuecomment-179755791

@DeepDiver1975 I think this should fix both cases.

Basically allow to set the fileId instead of a real node. The node is then only constructed on request.

CC: @schiesbn @PVince81 @nickvergessen

### CODE REVIEW

File: tests/lib/share20/defaultshareprovidertest.php [hunk 1]
```diff
@@ -171,6 +171,85 @@ public function testGetShareByIdUserShare() {
...
+		$this->assertEquals('sharedWith', $share->getSharedWith());
+		$this->assertEquals('sharedBy', $share->getSharedBy());
+		$this->assertEquals('shareOwner', $share->getShareOwner());
+		$this->assertEquals(13, $share->getPermissions());
+		$this->assertEquals(null, $share->getToken());
+		$this->assertEquals(null, $share->getExpirationDate());
+		$this->assertEquals('myTarget', $share->getTarget());
+	}
+
+	public function testGetShareByIdLazy2() {
```
Comment (DeepDiver1975) [inline tests/lib/share20/defaultshareprovidertest.php]: Can you please add a unit test for the delete case? THX

Comment (rullzer) [inline tests/lib/share20/defaultshareprovidertest.php]: sure thing.

Comment (rullzer) [inline tests/lib/share20/defaultshareprovidertest.php]: Done :)

**Comment by DeepDiver1975** [comment]
there is still an issue with delete ...

````
Restoring shares ...

                                 
  [OCP\Files\NotFoundException]  
                                 

Exception trace:
 () at /home/deepdiver/Development/ownCloud/master-autotest/lib/private/share20/share.php:123
 OC\Share20\Share->getNode() at /home/deepdiver/Development/ownCloud/master-autotest/lib/private/share20/manager.php:666
 OC\Share20\Manager->OC\Share20\{closure}() at /home/deepdiver/Development/ownCloud/master-autotest/lib/private/share20/manager.php:678
 OC\Share20\Manager->deleteShare() at /home/deepdiver/Development/ownCloud/master-autotest/apps/files/command/transferownership.php:216
 OCA\Files\Command\TransferOwnership->restoreShares() at /home/deepdiver/Development/ownCloud/master-autotest/apps/files/command/transferownership.php:118
 OCA\Files\Command\TransferOwnership->execute() at /home/deepdiver/Development/ownCloud/master-autotest/3rdparty/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /home/deepdiver/Development/ownCloud/master-autotest/3rdparty/symfony/console/Application.php:840
 Symfony\Component\Console\Application->doRunCommand() at /home/deepdiver/Development/ownCloud/master-autotest/3rdparty/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at /home/deepdiver/Development/ownCloud/master-autotest/3rdparty/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at /home/deepdiver/Development/ownCloud/master-autotest/lib/private/console/application.php:110
 OC\Console\Application->run() at /home/deepdiver/Development/ownCloud/master-autotest/console.php:85
 require_once() at /home/deepdiver/Development/ownCloud/master-autotest/occ:11

**Comment by DeepDiver1975** [comment]
@rullzer caused in here

https://github.com/owncloud/core/blob/share2_lazy_shares/lib/private/share20/manager.php#L666

**Comment by rullzer** [comment]
Aaaah stupid hooks... fix is on the way...

**Comment by rullzer** [comment]
@DeepDiver1975 please try again...

I also added a unit tests now for the manager... so all *should* work...

**Comment by DeepDiver1975** [comment]
THX - this really fixes both issues! :+1:

**Comment by PVince81** [comment]
Code looks good :+1:

**Pull request merged**